echidna/core/dsextract.py

```diff
@@ -0,0 +1,621 @@
+                                         (entry.posz)**2)) < self._fv_radius:
+                    return True  # Passes fiducial volume cut
+                return False  # Fails fiducial volume cut
+            return True  # Passes scintFit, has energy and no fid v cut
+        return False  # Fails scintFit or no energy
+
+    def get_value_ntuple(self, entry):
+        '''Get energy value from an ntuple EV
+
+        Args:
+          entry (:class:`ROOT.TChain`) chain entry
+
+        Returns:
+          float: Reconstructed energy
+        '''
+        return entry.energy
+
+
+class EnergyExtractTruth(Extractor):
+    '''True MC energy extraction methods.
+
+    Args:
+      fv_radius (float, optional): Fiducial radius. Applies a cut to remove
+        events which have a radial position greater than the radius of the
+        fiducial volume. If None no cut is applied.
+    '''
+
+    def __init__(self, fv_radius=None):
+        '''Initialise the class
+        '''
+        super(EnergyExtractTruth, self).__init__("energy_truth", fv_radius)
+
+    def get_valid_root(self, mc):
+        '''Check whether energy of a DS::MC is valid
+
+        Args:
+          mc (:class:`RAT.DS.MC`) entry
+
+        Returns:
+          bool: Validity boolean
+        '''
+        if mc.GetMCParticleCount > 0:
+            if self._fv_radius:
+                if mc.GetMCParticle(0).GetPosition().Mag() < self._fv_radius:
+                    return True  # Passes fiducial volume cut
+                return False  # Fails fiducial volume cut
+            return True  # Passes particle count and no fid v cut
+        return False  # Fails particle count
+
+    def get_value_root(self, mc):
+        '''Get energy value from a DS::MC
+
+        Args:
+          mc (:class:`RAT.DS.MC`) entry
+
+        Returns:
+          float: True energy
+        '''
+        return mc.GetScintEnergyDeposit()
+
+    def get_valid_ntuple(self, entry):
+        '''Check whether energy of an ntuple MC is valid
+
+        Args:
+          entry (:class:`ROOT.TChain`) chain entry
+
+        Returns:
+          bool: Validity boolean
+        '''
+        if entry.mc == 1:
+            if self._fv_radius:
+                if numpy.fabs(numpy.sqrt((entry.mcPosx)**2 +
+                                         (entry.mcPosy)**2 +
+                                         (entry.mcPosz)**2)) < self._fv_radius:
+                    return True  # Passes fiducial volume cut
+                return False  # Fails fiducial volume cut
+            # MC has associated trigger or vice versa and no fid v applied
+            return True
+        return False  # No associated trigger with mc or vice versa
+
+    def get_value_ntuple(self, entry):
+        '''Get energy value from an ntuple MC
+
+        Args:
+          entry (:class:`ROOT.TChain`) chain entry
+
+        Returns:
+          float: True energy
+        '''
+        entry.mcEdep
```

Comment:
@jwaterfield, found out why my Majoron spectra were coming out with blank histograms. Missing return statement here, so the extractor was extracting the correct energy, but then returning `None`.

Also there are a couple of pep coding errors in `fill_spectrum` that I just noticed whilst bug hunting, some undefined variables and unused imports, theses should be flagged up by running pep8 on the file. I'll point them out later.